Add golangci-lint configuration

imiric opened the pull request:
This adds a bunch of new linters based on the discussion below, and currently (on 50d0160) passes with `--new-from-rev` enabled. Without it there are 1298 issues... :sob:

See #58

imiric commented on .golangci.yml:
```diff
@@ -0,0 +1,39 @@
+    - structcheck
+    - unused
+    - varcheck
+    - staticcheck
```
Should we add more linters?

The [k6 config](https://github.com/grafana/k6/blob/0dda328d9a7f563607dae4fe89ff8c5138f8dc09/.golangci.yml) enables all and disables some, which I think is an error-prone approach, as new linters are constantly added, which often introduces new failures. So this is a variation of the [helm config](https://github.com/helm/helm/blob/35f6196167f6b90913f3a94237825856b858a215/.golangci.yml) which is a safer approach.

inancgumus replied:
Hey, I've summarized the other linters we might consider and the current number of warnings from them:

| linter | # of warnings |
| --- | --- |
| gocritic | 45 |
| errcheck | 3 |
| typecheck | 0 |
| cyclop | 37 |
| durationcheck | 1 |
| errchkjson | 6 |
| errorlint | 22 |
| exhaustive | 20 |
| exportloopref | 0 |
| forbidigo | 1 |
| funlen | 25 |
| gci | 33 |
| gochecknoglobals | 21 |
| gocognit | 11 |
| goconst | 33 |
| gocyclo | 1 |
| godot | 245 |
| gofumpt | 11 |
| gosec | 6 |
| ifshort | 6 |
| lll | 104 |
| makezero | 1 |
| misspell | 3 |
| nakedret | 0 |
| nestif | 11 |
| nilerr | 0 |
| nilnil | 4 |
| nlreturn | 435 |
| prealloc | 2 |
| thelper | 11 |
| tparallel | 6 |
| unconvert | 4 |
| unparam | 15 |
| wastedassign | 1 |
| whitespace | 4 |
| wrapcheck | 48 |
| forcetypeassert | 76 I'm not sure about this one. |

Should we enable `containedctx` or `ireturn` as well? Just kidding :)

inancgumus replied:
See: https://github.com/grafana/xk6-browser/pull/216#issuecomment-1029766734

inancgumus commented:
Hey, I've summarized the other linters we might consider and the current number of warnings from them:

| linter | # of warnings |
| --- | --- |
| gocritic | 45 |
| errcheck | 3 |
| typecheck | 0 |
| cyclop | 37 |
| durationcheck | 1 |
| errchkjson | 6 |
| errorlint | 22 |
| exhaustive | 20 |
| exportloopref | 0 |
| forbidigo | 1 |
| funlen | 25 |
| gci | 33 |
| gochecknoglobals | 21 |
| gocognit | 11 |
| goconst | 33 |
| gocyclo | 1 |
| godot | 245 |
| gofumpt | 11 |
| gosec | 6 |
| ifshort | 6 |
| lll | 104 |
| makezero | 1 |
| misspell | 3 |
| nakedret | 0 |
| nestif | 11 |
| nilerr | 0 |
| nilnil | 4 |
| nlreturn | 435 |
| prealloc | 2 |
| thelper | 11 |
| tparallel | 6 |
| unconvert | 4 |
| unparam | 15 |
| wastedassign | 1 |
| whitespace | 4 |
| wrapcheck | 48 |
| forcetypeassert | 76 I'm not sure about this one. |

Should we enable `containedctx` or `ireturn` as well? Just kidding :)

imiric commented:
Seems like a good list, but I'll have to take a closer look at each one. There might be redundant ones there, like `cyclop` which seems to cover all use cases of `gocyclo`.

But how do you want to approach this? Re-enable `--new-from-rev` in CI and run without it locally, while we gradually work on resolving the issues? Maybe some could be automatically resolved, or scripted.

Either way, this is not high priority, so we don't have to start on it now.

inancgumus commented:
> There might be redundant ones there

I believe others check for different stuff. You may have found the only overlapping one. But sure, take a closer look, that won't hurt, and I might be wrong 🤷 

> But how do you want to approach this?

Agreed, fixing these linter issues isn't a high priority. However, sometimes subtle bugs may have been caused because of these issues and can take more time. Re-enabling `--new-from-rev` seems like a better option, for now, so we can stop introducing more warnings.

I believe we better work on them one day/week for the remaining ones. So we can finish them in a month or so.

imiric commented:
@inancgumus I enabled most of the linters you listed, and added a few more. I'm concerned we're running way too many, though...

Some comments:
- I relaxed blocks for `nlreturn`, so it only triggers on >10 LOC.
- `staticcheck` seems to be a `go vet` replacement (from `golangci-lint help linters`: `Staticcheck is a go vet on steroids`), so can we disable `govet`?
- Added `importas`, `nolintlint`, `predeclared`, `tagliatelle`, and `varnamelen`. I'm not too sure if we should enable these last two ones, especially `varnamelen`, as it can get annoying. See https://github.com/grafana/k6/pull/2323.

I'll fix the existing errors after `--new-from-rev` in this PR and we can hopefully merge it then.

imiric commented on common/connection.go:
```diff
@@ -88,7 +88,7 @@ func (f ActionFunc) Do(ctx context.Context) error {
 │handler for a specific CDP │             │   Event Listener   │      *  *  *  *  *      │   Event Listener   │
 │       Domain event.       │             │                    │                         │                    │
-└───────────────────────────┘             └────────────────────┘                         └────────────────────┘
+└───────────────────────────┘             └────────────────────┘                         └────────────────────┘.
```
:man_facepalming:

inancgumus commented on common/frame.go:
```diff
@@ -1538,6 +1538,7 @@ type frameExecutionContext interface {
 	ID() runtime.ExecutionContextID
 }
 
+//nolint:unparam
```
I would rather keep this warning on and refactor these functions, also `framePointerActionFn`.

imiric replied:
Sure, but I didn't want to get into a large refactoring now. So we need to keep this `nolint`, and the `TODO` could be "Get rid of all `nolint`s".

inancgumus commented:
> @inancgumus I enabled most of the linters you listed, and added a few more. I'm concerned we're running way too many, though...

I prefer removing or adjusting them until we hit an issue. I diff'ed your changes, you actually added all the linters I suggested but gocyclo as we discussed 👍 

> I relaxed blocks for nlreturn, so it only triggers on >10 LOC.

👍 

> staticcheck seems to be a go vet replacement (from golangci-lint help linters: Staticcheck is a go vet on steroids), so can we disable govet?

We better keep it, see https://staticcheck.io/docs/: _"Staticcheck aims to be trivial to adopt. It behaves just like the official go tool and requires no learning to get started with. Just run staticcheck ./... on your code **in addition to go vet ./....**"_

> Added importas, nolintlint, predeclared, tagliatelle, and varnamelen. I'm not too sure if we should enable these last two ones, especially varnamelen, as it can get annoying. See https://github.com/grafana/k6/pull/2323.

These ones are ok but I'm strongly against varnamelen. It's too noisy and produced non-idiomatic warnings (about 1k).

> I'll fix the existing errors after --new-from-rev in this PR and we can hopefully merge it then.

Are you going to fix every error or do it as a team in a weekly fashion as we discussed?

inancgumus commented on .golangci.yml:
```diff
@@ -1,3 +1,6 @@
+# v1.44.0
+# Please don't remove the first line. It's used in CI to determine the
+# golangci-lint version.
```
```suggestion
# Please don't remove the first line. It's used in CI to determine the
# golangci-lint version. See: The lint step in .github/workflows/all.yaml
```

It's good to have this kind of verbose explanation for newcomers to our team.

inancgumus commented on common/element_handle.go:
```diff
@@ -106,6 +106,7 @@ func getElementHandleActionFn(
 	}
 }
 
+//nolint:funlen,gocognit,cyclop,unparam
```
We better refactor this one in TODO [here](https://github.com/grafana/xk6-browser/issues/58#issue-1044893666) instead of silencing it.

inancgumus replied:
Can you add a TODO to #58? Also for the other one.

imiric replied:
Just added a generic "fill all `nolint`s".

inancgumus commented on common/element_handle.go:
```diff
@@ -64,7 +64,7 @@ type (
 
 func getElementHandleActionFn(
-	h *ElementHandle, states []string, fn elementHandleActionFn, force, noWaitAfter bool, timeout time.Duration,
+	elh *ElementHandle, states []string, ehaFn elementHandleActionFn, force, noWaitAfter bool, timeout time.Duration,
```
I'm against varnamelen linter and on the side of reverting these changes. There is nothing wrong accepting `h` or `fn`. It is idiomatic to have these as short parameters. Their types make it clear what's going on.

imiric replied:
I'm not sure... I see the benefit of `varnamelen`, even in this case. The scope of this function is too large to use vars like `h` and `fn`. This is fine for ~5 LOC functions or `for` loops, but not here.

We definitely abuse this often, as you can see from the 339 errors (not sure how you got 1k?). But I think most of these can be quickly changed using `sed` or [`rf`](https://pkg.go.dev/rsc.io/rf). And we can always [tweak the configuration](https://github.com/blizzy78/varnamelen#golangci-lint-integration) to e.g. relax the scope requirements.

So I'm leaning towards keeping it, as it's valuable with something we're likely to not point out in PRs.

inancgumus replied:
Sorry, I see no benefit here, and I'm firmly against it. This is not something we want to enforce on senior devs. Better use linters for practical purposes instead of forcing non-idiomatic style changes.

inancgumus replied:
I just ran this useless linter on Go stdlib and it found 14K warnings. Come on.

inancgumus replied:
I just ran the linter on Go stdlib and it found 14K warnings. Come on.

imiric replied:
It's not an issue of seniority :smile:, but properly naming vars according to the scope they're used in. It's about legibility, avoiding shadowing, `grep`ability, etc.

I disagree, but if you feel strongly about it, sure, I'll remove it.

inancgumus replied:
Please remove it 🙏 

Btw, Rob won't agree. I ran the linter on his ivy project: 300 warnings. It seems he can't write idiomatic code according to this linter...

imiric replied:
Ah, I certainly wouldn't want to be maintaining his ivy project then :stuck_out_tongue:

FWIW, I prefer not blindly sticking to whatever is "idiomatic", but to whatever improves the DX while working on the code. To me it's a no brainer that longer var names _for larger scopes_ are the better approach.

inancgumus replied:
> longer var names for larger scopes are the better approach

I agree and but it depends on a blend of a dev's experience + code at hand. An automatic tool cannot decide that. When I recommend something, it's because I know it from my experience, not because I follow people blindly.

inancgumus reviewed (changes requested):
LGTM, nice work, and as usual, I have some suggestions.

imiric commented:
> Are you going to fix every error or do it as a team in a weekly fashion as we [discussed](https://github.com/grafana/xk6-browser/pull/216#issuecomment-1029797172)?

That's definitely still the plan :sweat_smile: I just had to fix/silence these initial ones as they were getting triggered with this configuration even with `--new-from-rev`.

inancgumus reviewed (approved):
LGTM 👏
